scala/lambdas/ingest-flow-control/src/test/scala/uk/gov/nationalarchives/ingestflowcontrol/LambdaTest.scala

```diff
@@ -0,0 +1,425 @@
+
+  def ssmClient(ref: Ref[IO, FlowControlConfig], errors: Option[Errors]): DASSMClient[IO] = new DASSMClient[IO]:
+    override def getParameter[T](parameterName: String, withDecryption: Boolean)(using Decoder[T]): IO[T] =
+      IO.whenA(errors.exists(_.getParameter))(IO.raiseError(new Exception("Error getting parameter"))) >> ref.get.map(_.asInstanceOf[T])
+
+  def dynamoClient(ref: Ref[IO, List[IngestQueueTableItem]]): DADynamoDBClient[IO] = new DADynamoDBClient[IO]:
+    override def deleteItems[T](tableName: String, primaryKeyAttributes: List[T])(using DynamoFormat[T]): IO[List[BatchWriteItemResponse]] = {
+      ref
+        .update { r =>
+          r.filterNot { row =>
+            primaryKeyAttributes.contains(IngestQueuePrimaryKey(IngestQueuePartitionKey(row.sourceSystem), IngestQueueSortKey(row.queuedAt)))
+          }
+        }
+        .map(_ => Nil)
+    }
+
+    override def writeItem(dynamoDbWriteRequest: DADynamoDBClient.DADynamoDbWriteItemRequest): IO[Int] = ref
+      .update { existing =>
+        IngestQueueTableItem(
+          dynamoDbWriteRequest.attributeNamesAndValuesToWrite(sourceSystem).s(),
+          Instant.parse(dynamoDbWriteRequest.attributeNamesAndValuesToWrite(queuedAt).s()),
+          dynamoDbWriteRequest.attributeNamesAndValuesToWrite(taskToken).s()
+        ) :: existing
+      }
+      .map(_ => 1)
+
+    override def writeItems[T](tableName: String, items: List[T])(using format: DynamoFormat[T]): IO[List[BatchWriteItemResponse]] = notImplemented
+    override def queryItems[U](tableName: String, requestCondition: RequestCondition, potentialGsiName: Option[String])(using returnTypeFormat: DynamoFormat[U]): IO[List[U]] =
+      notImplemented
+
+    override def getItems[T, K](primaryKeys: List[K], tableName: String)(using returnFormat: DynamoFormat[T], keyFormat: DynamoFormat[K]): IO[List[T]] =
+      val firstPK = primaryKeys.head.asInstanceOf[IngestQueuePartitionKey]
+      ref.get.map { existing =>
+        existing.filter(_.sourceSystem == firstPK.sourceSystem).sortBy(_.queuedAt).map(_.asInstanceOf[T])
+      }
+
+    override def updateAttributeValues(dynamoDbRequest: DADynamoDBClient.DADynamoDbRequest): IO[Int] = notImplemented
+
+  def sfnClient(ref: Ref[IO, List[StepFunctionExecution]]): DASFNClient[IO] = new DASFNClient[IO]:
+    override def startExecution[T <: Product](stateMachineArn: String, input: T, name: Option[String])(using enc: Encoder[T]): IO[StartExecutionResponse] = notImplemented
+
+    override def listStepFunctions(stepFunctionArn: String, status: DASFNClient.Status): IO[List[String]] = ref.get.map { existing =>
+      existing.map(_.name)
+    }
+
+    override def sendTaskSuccess(token: String): IO[Unit] = {
+      ref
+        .update { existing =>
+          val updatedExecution = existing.filter(_.taskToken == token).map(_.copy(taskTokenSuccess = true))
+          existing.filter(_.taskToken != token) ++ updatedExecution
+        }
+    }
+
+  "lambda" should "do stuff" in {
```

Comment:
😄 , maybe a bit too vague? how about... "do some really cool stuff"? But honestly, how come it returns errors?

Reply:
arghhh, that was the first "template" test I had written, it was for the error case for SSM, I've updated the name now.